autoconnect/autoconnect-ws/autoconnect-ws-sm/src/identified/mod.rs

```diff
@@ -82,35 +98,140 @@ impl WebPushClient {
+            "ua_browser_family" => &ua_info.metrics_browser,
+            "ua_browser_ver" => &ua_info.browser_version,
+            "ua_category" => &ua_info.category,
+            "connection_time" => elapsed_sec,
+            "direct_acked" => stats.direct_acked,
+            "direct_storage" => stats.direct_storage,
+            "stored_retrieved" => stats.stored_retrieved,
+            "stored_acked" => stats.stored_acked,
+            "nacks" => stats.nacks,
+            "registers" => stats.registers,
+            "unregisters" => stats.unregisters,
+            "disconnect_reason" => reason.unwrap_or_else(|| "".to_owned()),
+        );
+    }
+
+    /// Save any Direct unAck'd messages to the db (on shutdown)
+    ///
+    /// Direct messages are solely stored in memory until Ack'd by the Client,
+    /// so on shutdown, any not Ack'd are stored in the db to not be lost
+    fn save_and_notify_unacked_direct_notifs(&mut self) {
+        trace!("WebPushClient::save_and_notify_unacked_direct_notifs");
+        let mut notifs = mem::take(&mut self.ack_state.unacked_direct_notifs);
+        self.stats.direct_storage += notifs.len() as i32;
+        // TODO: clarify this comment re the Python version
+        // Ensure we don't store these as legacy by setting a 0 as the
+        // sortkey_timestamp. This ensures the Python side doesn't mark it as
+        // legacy during conversion and still get the correct default us_time
+        // when saving
+        for notif in &mut notifs {
+            notif.sortkey_timestamp = Some(0);
+        }
+
+        let app_state = Arc::clone(&self.app_state);
+        let uaid = self.uaid;
+        let connected_at = self.connected_at;
+        rt::spawn(async move {
+            app_state.db.save_messages(&uaid, notifs).await?;
+            debug!("Finished saving unacked direct notifs, checking for reconnect");
+            let Some(user) = app_state.db.get_user(&uaid).await? else {
+                return Err(SMError::Internal(format!("User not found for unacked direct notifs: {uaid}")));
+            };
+            if connected_at == user.connected_at {
+                return Ok(());
+            }
+            if let Some(node_id) = user.node_id {
+                app_state
+                    .http
+                    .put(&format!("{}/notif/{}", node_id, uaid.as_simple()))
+                    .send()
+                    .await?
+                    .error_for_status()?;
+            }
+            Ok(())
+        });
     }
 }
 
-#[allow(dead_code)]
+/// Ensure an existing user's record is valid, returning its `ClientFlags`
+///
+/// Somewhat similar to autoendpoint's `validate_webpush_user` function. When a
+/// User record is invalid it will be dropped from the db and `None` will be
+/// returned.
+pub async fn process_existing_user(
+    app_state: &Arc<AppState>,
+    user: &User,
+) -> DbResult<Option<ClientFlags>> {
+    if user.current_month.is_none() {
+        app_state
```

Comment:
Per our discussion, just to clarify: we aren't hitting `current_month.is_none()` We do see similar metrics but they're for legacy autoconnect dropping users due to their message month being outside our list of "valid" message tables.

We can see we still have more than one message month table (probably 1 or 2 behind prod's current month) because with users still in those months because they're continuing to trigger the message table rotation logic on legacy autoconnect.